PrestaShop/PrestaShop
# Rework OrderHandlers calculations to use Prestashop Number

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | See below
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | part of #17282 and #18020
| How to test?  | See below

## Description

### Initial scope

- fix `@todo` in **AddCartRuleToOrderHandler** ('move to private method' in couple places)

### Scope expansion

- refactors **AddCartRuleToOrderHandler**, **AddProductToOrderHandler**, **DeleteProductFromOrderHandler**, and **UpdateProductInOrderHandler** handlers into private methods
- implement Number calculations instead of floats for handlers **AddCartRuleToOrderHandler**, **AddProductToOrderHandler**, **DeleteProductFromOrderHandler**, and **UpdateProductInOrderHandler**
- added some behat tests
- removed the ability to update invoices when adding discount to order _as requested by PM_
- fixed bug where "free shipping" checkbox wasn't displayed when adding product with new invoice

## How to test ?

Use `sell/orders/orders/[orderId]/view` to reach migrated order view page. Make sure `Add discount` is working as expected.
**Tip:** to add product with new invoice the order status must be "payment accepted" (works with some others as well).
Be sure to check the product total prices (shipping, total, taxes, discounts- _this one also increments when product is added with free shipping option_).
:warning:  **WARNING**
Pr included a heavy rebase associated with price calculations from #17877 and #17476. So its important to test according to those PR's as well.

## Known issues
* when adding product with free shipping, the discounts are increased, but not shipping price, which leads to wrong total numbers. (the overall total for the order is correct though). _(this one is hard to fix and is delayed to at least 1.7.8)_
* After adding product with new invoice and free shipping, the page needs to be refreshed to see the new specific price rule in list. _(this should be solved in another PR)_


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/17348)
